pl_bolts/datamodules/experience_source.py

```diff
@@ -0,0 +1,201 @@
+"""
+Datamodules for RL models that rely on experiences generated during training
+
+Based on implementations found here: https://github.com/Shmuma/ptan/blob/master/ptan/experience.py
+"""
+from collections import deque
+from typing import Iterable, Callable, Tuple, List
+import numpy as np
+import torch
+from gym import Env
+from torch.utils.data import IterableDataset
+
+# Datasets
+from pl_bolts.models.rl.common.agents import Agent
+from pl_bolts.models.rl.common.memory import Experience
+
+
+class ExperienceSourceDataset(IterableDataset):
+    """
+    Basic experience source dataset. Takes a generate_batch function that returns an iterator.
+    The logic for the experience source and how the batch is generated is defined the Lightning model itself
+    """
+
+    def __init__(self, generate_batch: Callable):
+        self.generate_batch = generate_batch
+
+    def __iter__(self) -> Iterable:
+        iterator = self.generate_batch()
+        return iterator
+
+# Experience Sources
+
+
+class ExperienceSource:
+    """
+    Basic single step experience source
+
+    Args:
+        env: Environment that is being used
+        agent: Agent being used to make decisions
+    """
+
+    def __init__(self, env: Env, agent: Agent):
+        self.env = env
+        self.agent = agent
+        self.state = self.env.reset()
+
+    def _reset(self) -> None:
+        """resets the env and state"""
+        self.state = self.env.reset()
+
+    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
+        """Takes a single step through the environment"""
+        action = self.agent(self.state, device)
+        new_state, reward, done, _ = self.env.step(action)
+        experience = Experience(
+            state=self.state,
+            action=action,
+            reward=reward,
+            new_state=new_state,
+            done=done,
+        )
+        self.state = new_state
+
+        if done:
+            self.state = self.env.reset()
+
+        return experience, reward, done
+
+    def run_episode(self, device: torch.device) -> float:
+        """Carries out a single episode and returns the total reward. This is used for testing"""
+        done = False
+        total_reward = 0
+
+        while not done:
+            _, reward, done = self.step(device)
+            total_reward += reward
+
+        return total_reward
+
+
+class NStepExperienceSource(ExperienceSource):
+    """Expands upon the basic ExperienceSource by collecting experience across N steps"""
+
+    def __init__(self, env: Env, agent: Agent, n_steps: int = 1, gamma: float = 0.99):
+        super().__init__(env, agent)
+        self.gamma = gamma
+        self.n_steps = n_steps
+        self.n_step_buffer = deque(maxlen=n_steps)
+
+    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
```

Comment:
Can we change it to make this one a single step (as in the base class) for consistency and instead of single_step add a  n-step method?

Reply:
Yeah no problem